File: Samples/DeviceEnumerationAndPairing/cppwinrt/Scenario5_GetSingleDevice.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario5_GetSingleDevice.h"
#include "SampleConfiguration.h"
#include "DeviceHelpers.h"
#include "Scenario5_GetSingleDevice.g.cpp"

using namespace winrt;
using namespace winrt::Windows::Devices::Enumeration;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::UI::Xaml;
using namespace winrt::Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario5_GetSingleDevice::Scenario5_GetSingleDevice()
    {
        InitializeComponent();
    }

    void Scenario5_GetSingleDevice::OnNavigatedTo(NavigationEventArgs const&)
    {
        resultsListView().ItemsSource(resultCollection);
    }

    fire_and_forget Scenario5_GetSingleDevice::Page_Loaded(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        // Pre-populate the text box with an interface id.
         DeviceInformationCollection deviceInfoCollection = co_await DeviceInformation::FindAllAsync(DeviceClass::AudioRender);
        if (deviceInfoCollection.Size() > 0)
        {
            // When you want to "save" a DeviceInformation to get it back again later,
            // use both the DeviceInformation.Kind and the DeviceInformation.Id.
            DeviceInformation info = deviceInfoCollection.GetAt(0);
            interfaceIdTextBox().Text(info.Id());
            deviceInformationKind = info.Kind();
            InformationKindTextBox().Text(to_hstring(deviceInformationKind));
            getButton().IsEnabled(true);
        }
    }

    fire_and_forget Scenario5_GetSingleDevice::GetButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();
        resultCollection.Clear();

        hstring interfaceId = interfaceIdTextBox().Text();
        getButton().IsEnabled(false);

        rootPage.NotifyUser(L"CreateFromIdAsync operation started...", NotifyType::StatusMessage);

        try
        {
            // When you want to "save" a DeviceInformation to get it back again later,
            // use both the DeviceInformation.Kind and the DeviceInformation.Id.
            DeviceInformation deviceInfo = co_await DeviceInformation::CreateFromIdAsync(interfaceId, nullptr, deviceInformationKind);

            rootPage.NotifyUser(L"CreateFromIdAsync operation completed.", NotifyType::StatusMessage);

            resultCollection.Append(make<DeviceInformationDisplay>(deviceInfo));
        }
        catch (hresult_error const& e)
        {
            if (e.code() == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND))
            {
                rootPage.NotifyUser(L"Invalid Interface Id", NotifyType::ErrorMessage);
            }
            else
            {
                throw;
            }
        }

        getButton().IsEnabled(true);
        getButton().Focus(FocusState::Keyboard);
    }
}

File: Samples/DeviceEnumerationAndPairing/js/js/scenario5_getsingledevice.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var DevEnum = Windows.Devices.Enumeration;
    var resultCollection = new WinJS.Binding.List([]);
    var resultsListView;
    var interfaceIdTextBox;
    var getButton;
    var deviceInformationKind;

    var page = WinJS.UI.Pages.define("/html/scenario5_getsingledevice.html", {
        ready: function (element, options) {

            resultsListView = element.querySelector("#resultsListView").winControl;
            interfaceIdTextBox = document.getElementById("interfaceIdTextBox");
            getButton = document.getElementById("getButton");

            // Hook up button event handlers
            getButton.addEventListener("click", getButtonClick, false);

            // Hook up result list data binding
            resultsListView.itemDataSource = resultCollection.dataSource;

            // Capture this in case the user navigates before the findAllAsync completes.
            var informationKindTextBox = document.getElementById("informationKindTextBox");

            // Pre-populate the interface id input box with something valid.
            DevEnum.DeviceInformation.findAllAsync(DevEnum.DeviceClass.audioRender).done(
                function (deviceInfoCollection) {
                    if (deviceInfoCollection.length > 0) {
                        // When you want to "save" a DeviceInformation to get it back again later,
                        // use both the DeviceInformation.Kind and the DeviceInformation.Id.
                        interfaceIdTextBox.value = deviceInfoCollection.getAt(0).id;
                        deviceInformationKind = deviceInfoCollection.getAt(0).kind;
                        informationKindTextBox.value = DisplayHelpers.kindStringFromType(deviceInformationKind);
                        getButton.disabled = false;
                    }
                });

            // Process any data bindings
            WinJS.UI.processAll();
        },
        unload: function () {
            resultCollection.splice(0, resultCollection.length);
        }
    });

    function getButtonClick() {

        getButton.disabled = true;
        resultCollection.splice(0, resultCollection.length);
        
        WinJS.log && WinJS.log("CreateFromIdAsync operation started...", "sample", "status");

        // When you want to "save" a DeviceInformation to get it back again later,
        // use both the DeviceInformation.Kind and the DeviceInformation.Id.
        DevEnum.DeviceInformation.createFromIdAsync(interfaceIdTextBox.value, null, deviceInformationKind).then(
            function (deviceInfo) {
                WinJS.log && WinJS.log("CreateFromIdAsync operation completed.", "sample", "status");
                resultCollection.push(new DisplayHelpers.DeviceInformationDisplay(deviceInfo));
            },
            function (e) {
                if (e.number === (0x80070002 | 0)) { // HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND)
                    WinJS.log && WinJS.log("Invalid Interface Id", "sample", "error");
                } else {
                    throw e;
                }
            }).done(function () {
                getButton.disabled = false;
            });
    }

})();
